dora/core/common/src/main/java/alluxio/membership/EtcdMembershipManager.java

```diff
@@ -116,84 +115,126 @@ public void join(WorkerInfo workerInfo) throws IOException {
+    protected Stream<WorkerServiceEntity> getWorkersInView() {
+      return mAlluxioEtcdClient.getChildren(getRingPathPrefix())
+          .stream()
+          .map(super::decode)
+          .filter(Optional::isPresent)
+          .map(Optional::get);
     }
-    return fullMembers;
   }
 
-  private List<WorkerServiceEntity> retrieveLiveMembers() throws IOException {
-    List<WorkerServiceEntity> liveMembers = new ArrayList<>();
-    for (Map.Entry<String, ByteBuffer> entry : mAlluxioEtcdClient.mServiceDiscovery
-        .getAllLiveServices().entrySet()) {
-      try {
-        WorkerServiceEntity entity = new WorkerServiceEntity();
-        entity.deserialize(entry.getValue().array());
-        liveMembers.add(entity);
-      } catch (JsonSyntaxException ex) {
-        // Ignore
-      }
+  class LiveWorkersClusterView extends EtcdWorkerClusterView {
+    @Override
+    protected Stream<WorkerServiceEntity> getWorkersInView() {
+      return mAlluxioEtcdClient.mServiceDiscovery
+          .getAllLiveServices()
+          .stream()
+          .map(super::decode)
+          .filter(Optional::isPresent)
+          .map(Optional::get);
     }
-    return liveMembers;
   }
 
-  @Override
-  @VisibleForTesting
-  public List<WorkerInfo> getLiveMembers() throws IOException {
-    List<WorkerServiceEntity> liveWorkers = retrieveLiveMembers();
-    return liveWorkers.stream()
-        .map(e -> new WorkerInfo().setIdentity(e.getIdentity())
-            .setAddress(e.getWorkerNetAddress()))
-        .collect(Collectors.toList());
+  static class FailedWorkersClusterView extends EtcdWorkerClusterView {
+    private final AllWorkersClusterView mAllWorkers;
+    private final LiveWorkersClusterView mLiveWorkers;
+
+    private FailedWorkersClusterView(
+        AllWorkersClusterView allWorkers,
+        LiveWorkersClusterView liveWorkers) {
+      mAllWorkers = allWorkers;
+      mLiveWorkers = liveWorkers;
+    }
+
+    @Override
+    protected Stream<WorkerServiceEntity> getWorkersInView() {
+      Set<WorkerIdentity> liveWorkerIds = mLiveWorkers.getWorkersInView()
+          .map(WorkerServiceEntity::getIdentity)
+          .collect(Collectors.toSet());
+      return mAllWorkers.getWorkersInView()
+          .filter(w -> !liveWorkerIds.contains(w.getIdentity()));
+    }
   }
 
-  @Override
-  @VisibleForTesting
-  public List<WorkerInfo> getFailedMembers() throws IOException {
-    List<WorkerServiceEntity> registeredWorkers = retrieveFullMembers();
-    List<String> liveWorkers = retrieveLiveMembers()
-        .stream().map(e -> e.getServiceEntityName())
-        .collect(Collectors.toList());
-    registeredWorkers.removeIf(e -> liveWorkers.contains(e.getServiceEntityName()));
-    return registeredWorkers.stream()
-        .map(e -> new WorkerInfo().setIdentity(e.getIdentity())
-            .setAddress(e.getWorkerNetAddress()))
-        .collect(Collectors.toList());
+  abstract static class EtcdWorkerClusterView implements WorkerClusterView {
```

Comment:
why do we need 3 layer of hierarchy implementation for cluster view? 
why can't live/all/failed all be initialized with a WorkerClusterView  but with different list of workers?

Reply:
removed the class hierarchy.